File: src/python/espressomd/io/writer/h5md.py
#
# Copyright (C) 2010-2026 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import sys
import pathlib

from ...script_interface import script_interface_register, ScriptInterfaceHelper  # pylint: disable=import


class UnitSystem:
    """
    Data class for writing H5MD trajectories with
    `physical units <https://nongnu.org/h5md/modules/units.html>`__.
    There are four settable units: 'mass', 'length', 'time', 'charge'.
    Units should be written as strings following the specifications defined
    `here <https://nongnu.org/h5md/modules/units.html#unit-string>`__,
    e.g. ``UnitSystem(time='ps', mass='u', length='nm', charge='e')``.
    """

    def __init__(self, **kwargs):
        self.mass = ''
        self.time = ''
        self.length = ''
        self.charge = ''
        for key, value in kwargs.items():
            assert hasattr(self, key), f'unknown dimension {key}'
            setattr(self, key, value or '')

        if self.length and self.mass and self.time:
            self.force = f'{self.length} {self.mass} {self.time}-2'
        else:
            self.force = ''
        if self.length and self.time:
            self.velocity = f'{self.length} {self.time}-1'
        else:
            self.velocity = ''


@script_interface_register
class H5md(ScriptInterfaceHelper):

    """
    H5md file object.

    .. note::
       Bonds will be written to the file if they exist.
       The pypresso script will be written in the metadata.

    Parameters
    ----------
    file_path : :obj:`str` or :obj:`pathlib.Path`
        Path to the trajectory file, or an existing file to append data to
        (it must have the same specifications).
    unit_system : :obj:`UnitSystem`, optional
        Physical units for the data.
    fields : :obj:`set` or :obj:`str`, optional
        List of fields to write to the trajectory file. Defaults to ``'all'``.
        See :meth:`~espressomd.io.writer.h5md.H5md.valid_fields()` for the
        list of valid fields. This list defines the H5MD specifications.
        If the file in ``file_path`` already exists but has different
        specifications, an exception is raised.
    chunk_size : :obj:`int`
        The chunk size for hdf5 write operations. Must be greater than 0.

    Methods
    -------
    get_params()
        Get the parameters from the script interface.

    valid_fields()
        Get the list of valid fields.

    write()
        Call the H5md write method.

    flush()
        Call the H5md flush method. Only the dataset buffers will be flushed.
        The file superblock, which contains the datasets and metadata address
        space information (End of Allocation, EOA), won't be flushed. Only the
        H5md close method can flush the superblock. A file that was manually
        flushed via this method but not properly closed might have a stale EOA
        and thus be unreadable. This can happen when a simulation receives a
        termination signal and doesn't have time to release all file handles.

    close()
        Close the H5md file and flush both the data and superblock.

    Attributes
    ----------
    file_path: :obj:`str`
        Path to the trajectory file.
    script_path: :obj:`str`
        Path to the pypresso script, or empty string for interactive sessions.
    fields: :obj:`list`
        List of fields to write to the trajectory file.
    mass_unit: :obj:`str`
    length_unit: :obj:`str`
    time_unit: :obj:`str`
    force_unit: :obj:`str`
    velocity_unit: :obj:`str`
    charge_unit: :obj:`str`
    chunk_size: :obj:`int`

    """
    _so_name = "ScriptInterface::Writer::H5md"
    _so_features = ("H5MD",)
    _so_creation_policy = "GLOBAL"
    _so_bind_methods = ("valid_fields", "write", "flush", "close")

    def __init__(self, **kwargs):
        if "sip" in kwargs:
            super().__init__(**kwargs)
            return

        params = self.default_params()
        params.update(kwargs)
        unit_system = params.pop("unit_system")
        if isinstance(params["fields"], str):
            params["fields"] = [params["fields"]]
        script_path = ""
        if sys.argv and sys.argv[0]:
            script_path = str(pathlib.Path(sys.argv[0]).resolve())
        if not isinstance(params["file_path"], (str, pathlib.Path)):
            raise TypeError(
                "Parameter 'file_path' should be a string or a path")
        params["file_path"] = str(pathlib.Path(params["file_path"]).resolve())
        super().__init__(
            script_path=script_path,
            mass_unit=unit_system.mass,
            length_unit=unit_system.length,
            time_unit=unit_system.time,
            force_unit=unit_system.force,
            velocity_unit=unit_system.velocity,
            charge_unit=unit_system.charge,
            **params
        )

    def default_params(self):
        return {"unit_system": UnitSystem(), "fields": "all",
                "chunk_size": 1000}

    def required_keys(self):
        return {"file_path"}

    def valid_keys(self):
        return {"file_path", "unit_system", "fields", "chunk_size"}

File: src/script_interface/h5md/h5md.cpp
/*
 * Copyright (C) 2010-2026 The ESPResSo project
 * Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
 *   Max-Planck-Institute for Polymer Research, Theory Group
 *
 * This file is part of ESPResSo.
 *
 * ESPResSo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ESPResSo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config/config.hpp"

#ifdef ESPRESSO_H5MD

#include "h5md.hpp"

#include "core/cell_system/CellStructure.hpp"
#include "core/communication.hpp"
#include "core/io/writer/h5md_core.hpp"
#include "core/system/System.hpp"

#include <cassert>
#include <cmath>
#include <filesystem>
#include <string>
#include <vector>

namespace ScriptInterface {
namespace Writer {

H5md::H5md() {
  add_parameters(
      {{"file_path", m_h5md, &::Writer::H5md::File::file_path},
       {"chunk_size", m_h5md, &::Writer::H5md::File::chunk_size},
       {"script_path", m_h5md, &::Writer::H5md::File::script_path},
       {"fields", AutoParameter::read_only,
        [this]() { return make_vector_of_variants(m_output_fields); }},
       {"mass_unit", m_h5md, &::Writer::H5md::File::mass_unit},
       {"length_unit", m_h5md, &::Writer::H5md::File::length_unit},
       {"time_unit", m_h5md, &::Writer::H5md::File::time_unit},
       {"force_unit", m_h5md, &::Writer::H5md::File::force_unit},
       {"velocity_unit", m_h5md, &::Writer::H5md::File::velocity_unit},
       {"charge_unit", m_h5md, &::Writer::H5md::File::charge_unit}});
};

void H5md::do_construct(VariantMap const &params) {
  m_output_fields = get_value<std::vector<std::string>>(params, "fields");
  m_h5md =
      make_shared_from_args<::Writer::H5md::File, std::filesystem::path,
                            std::filesystem::path, std::vector<std::string>,
                            std::string, std::string, std::string, std::string,
                            std::string, std::string, int>(
          params, "file_path", "script_path", "fields", "mass_unit",
          "length_unit", "time_unit", "force_unit", "velocity_unit",
          "charge_unit", "chunk_size");
  // MPI communicator is needed to close parallel file handles
  m_mpi_env_lock = ::communication_environment->get_mpi_env();
  is_file_closed = false;
}

H5md::~H5md() {
  m_h5md.reset();
  assert(m_h5md.use_count() == 0u);
  m_mpi_env_lock.reset();
}

void H5md::sanity_check_is_file_closed(std::string const &name) const {
  if (is_file_closed) {
    if (context()->is_head_node()) {
      throw std::runtime_error("cannot call '" + name + "' on a closed file");
    }
    throw Exception("");
  }
}

Variant H5md::do_call_method(std::string const &name, VariantMap const &) {
  if (name == "write") {
    sanity_check_is_file_closed(name);
    auto const &system = ::System::get_system();
    // Forces/torques live in the ParticleStore columns, which are only rebuilt
    // lazily. A write triggered before the first integrator step (or after a
    // particle was added) would otherwise read stale/detached rows and crash;
    // sync the store first.
    system.cell_structure->ensure_particle_store_synchronized();
    auto const particles = system.cell_structure->local_particles();
    auto const sim_time = system.get_sim_time();
    auto const time_step = system.get_time_step();
    auto const n_steps = static_cast<int>(std::round(sim_time / time_step));
    m_h5md->write(particles, sim_time, n_steps, *system.box_geo);
  } else if (name == "flush") {
    sanity_check_is_file_closed(name);
    m_h5md->flush();
  } else if (name == "close") {
    sanity_check_is_file_closed(name);
    m_h5md->close();
    is_file_closed = true;
  } else if (name == "valid_fields") {
    return make_vector_of_variants(m_h5md->valid_fields());
  }
  return {};
}

} // namespace Writer
} // namespace ScriptInterface

#endif // ESPRESSO_H5MD
